### DISCUSSION TITLE
Quarkus Extension as Part of a Project inside a Monorepo?

### DISCUSSION THREAD

**Discussion by skurfuerst** [quarkusio/quarkus#24709] [Q&A] [answered]
Hey everybody,

I am currently getting my feet wet with Quarkus for the first time - and it is really a great product with an awesome dev experience.

One thing where I am currently struggling is that I'd love to experiment more with Quarkus extensions and the build-time code analysis capabilities. However, I'd love a tight dev loop, almost as tight as I get it running `quarkus dev` - that's why, if possible, **I'd love to have the extension's code as part of my project as well, to be able to access the compile time code generation stuff.**

Is there any way to make this work? In the docs, it says "you need to publish your extension to the local maven registry before using it", but I feel this would very much complicate my use case.

Thanks for any hints and all the best :) 
Sebastian

**Comment by maxandersen** [comment]
You can't unfortunately develop extensions with same speed as apps on top extensions if that is what you are asking for ?

**Reply by skurfuerst** [reply]
Hey,

Thanks for your quick reply!

So I would be interested in two things:
- is it possible to add some kind of deployment module to the current quarkus app? Like putting deployment sources into a new source directory or so :) it is clear for me that code changes in there need a full rebuild 
- how are others developing quarkus extensions workflow wise with a manageable dev roundtrip time?

Thanks so much :)

All the best,
Sebastian

**Reply by maxandersen** [reply]
I'm actually on PTO so can't digg more atm. Maybe @aloubyansky has some tips?

**Comment by aloubyansky** [comment, marked as answer]
Extension modules can be included in the codebase with the app. But live reload won't be working for extension modules and their deps, there will be warning logged about that.

**Reply by skurfuerst** [reply]
Hey @aloubyansky,

awesome, that's what I am trying to archive. Somehow it did not work for me with multiple modules - do you have more hints (or some project example at github) what I need to do? I'd be grateful for any hints or examples :)

All the best,
Sebastian

**Reply by aloubyansky** [reply]
Sure, here is a template used in the Quarkus testsuite https://github.com/quarkusio/quarkus/tree/main/integration-tests/maven/src/test/resources-filtered/projects/project-with-extension

**Reply by skurfuerst** [reply]
Perfect, I'll try this out tomorrow and give feedback :):)

**Reply by skurfuerst** [reply]
Hey,

I've tried it out now completely, and after some fiddling and further replacements, `project-with-extension` works now. Attached is a minimal example:
[project-with-extension.zip](https://github.com/quarkusio/quarkus/files/8412977/project-with-extension.zip)

To run this, you can NOT use `quarkus dev`,  but you need to do `mvn package quarkus:dev` - but then it works without problems :) And the dev round trip time is pretty quick too (you need a restart the mvn command if you change anything in the quarkus extension, but that is to be expected).

Thank you so much and keep up the great work :) 
Sebastian

**Reply by kpagratis** [reply]
Hello, I'm wondering if this answer is still valid in the latest versions of quarkus. Also, has anyone been able to make this work with gradle?

**Reply by aloubyansky** [reply]
It is still valid. It should work in the same way in Gradle.
We have projects with extensions in our testsuite, for example https://github.com/quarkusio/quarkus/tree/main/integration-tests/gradle/src/main/resources/basic-composite-build-extension-project. Perhaps that could help to create a simple project to test it. It's missing some version properties that are filled in during the test setup but it should given an idea.

**Reply by nort3x** [reply]
Hi, @aloubyansky 
this is composite build (included build) and it's almost always used when we want to include an external project and not really a submodule!

i recently configured a monorepo pretty much like the example you provided but i couldn't include the extension in my consumer service using regular `implementation(project(":library:extension-A"))`

is there any resource that i can look into why this is not possible?

**Reply by aloubyansky** [reply]
@nort3x would you be able to share a reproducer? That could make it easier to investigate and create a proper test. Thanks.

**Reply by nort3x** [reply]
here is an example, containing 1 quarkus extension (called quarkus-base)
one generic library (common-kotlin)
one service (serivce-a)

service-a uses both of the libraries,
`quarkus-base` should configure jackson for service-a and provide a `FeatureBuildItem` (both will produce log)
i'm migrating from includedBuilds to direct includes and i can't resolve this,
i'm also providing testFixture for service-a through quarkus-base (for instance some configuration for restassured)



[sample.zip](https://github.com/user-attachments/files/18705323/sample.zip)

**Reply by aloubyansky** [reply]
@nort3x could you please also create a github issue for it? Thanks.

**Reply by nort3x** [reply]
@aloubyansky [done](https://github.com/quarkusio/quarkus/issues/46137)

**Reply by nort3x** [reply]
@aloubyansky  i'm willing to invest time in this, any guide for contrib? or maybe another medium for communication?

**Reply by aloubyansky** [reply]
@nort3x sorry for the delay. I haven't had the time to dig into that, but the issue and the reproducer are very much appreciated.

W/o looking deeper, I think you'd have to look into how the `ApplicationModel` is resolved and initialized, which unfortunately is quite complicated. It'd be https://github.com/quarkusio/quarkus/blob/main/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/GradleApplicationModelBuilder.java
and the configurations are setup in https://github.com/quarkusio/quarkus/blob/main/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java

I've been looking into re-working the way we setup configurations but there are still a few issues I have to figure out.

The other channel would be https://quarkusio.zulipchat.com/

**Reply by nort3x** [reply]
@aloubyansky 
thank you, I'm willing to help and contrib plus due my android background i have a decent gradle knowledge.
i will look into provided resources but from the looks of it it is very internal and not well document so i don't think i can fix this myself  .
